{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "115bb5c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import gffutils\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccffe49d-7787-4631-906a-709b390d4f93",
   "metadata": {},
   "source": [
    "### Unzip db builds (if necessary):\n",
    "\n",
    "    tar -zxf nextflow_db.tar.gz\n",
    "    tar -zxf midasdb_localdb.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4af8ee08",
   "metadata": {},
   "source": [
    "### Output of nextflow pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "71b0f0c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_db_path = '/wynton/protected/home/sirota/clairedubin/MIDAS3_nextflow/testing/nextflow_db'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44904b08",
   "metadata": {},
   "source": [
    "### Output run generated Chunyu using original MIDAS scripts\n",
    "\n",
    "Original code here: https://github.com/czbiohub-sf/MIDAS/blob/master/tests/test_db.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b7abf987",
   "metadata": {},
   "outputs": [],
   "source": [
    "correct_db_path = '/wynton/protected/home/sirota/clairedubin/MIDAS3_nextflow/testing/midasdb_localdb'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3893d82",
   "metadata": {},
   "source": [
    "### Comparing folders\n",
    "\n",
    "Mine is missing genomes.tsv, markers_models, and cleaned_imports because the nextflow pipeline doesn't require these input files to be stored within the database directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "79d32421",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chunks\tgene_annotations  markers  pangenomes  pangenomes_annotation\n"
     ]
    }
   ],
   "source": [
    "!ls {my_db_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fd14dc2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chunks\t\t gene_annotations  markers\t   pangenomes\n",
      "cleaned_imports  genomes.tsv\t   markers_models  pangenomes_annotation\n"
     ]
    }
   ],
   "source": [
    "!ls {correct_db_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b563a780",
   "metadata": {},
   "source": [
    "### Get list of files in each database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2acbd0dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = 'nextflow_db'\n",
    "my_db_files = []\n",
    "\n",
    "for path, subdirs, files in os.walk(db):\n",
    "    for name in files:\n",
    "        \n",
    "        if name == 'genomes.tsv' or 'cleaned_imports' in path:\n",
    "            continue\n",
    "        \n",
    "        my_db_files += [os.path.join(path, name).replace(db+'/', '')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6bf9a8a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = 'midasdb_localdb'\n",
    "correct_db_files = []\n",
    "\n",
    "for path, subdirs, files in os.walk(db):\n",
    "    for name in files:\n",
    "        if '.log' in name: #ignore missing log files\n",
    "            continue\n",
    "        if name == 'genomes.tsv' or 'cleaned_imports' in path or 'markers_models' in path:\n",
    "            continue\n",
    "        correct_db_files += [os.path.join(path, name).replace(db+'/', '')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab7219e9",
   "metadata": {},
   "source": [
    "### Compare list of files generated by each database\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "250bc825",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pangenomes/117088/temp/vsearch/genes.ffn\n",
      "possible matches:\n",
      "pangenomes/117088/genes.ffn\n",
      "pangenomes/117088/temp/cdhit/genes.ffn\n",
      "\n",
      "pangenomes/117086/temp/vsearch/genes.ffn\n",
      "possible matches:\n",
      "pangenomes/117086/genes.ffn\n",
      "pangenomes/117086/temp/cdhit/genes.ffn\n",
      "\n"
     ]
    }
   ],
   "source": [
    "not_matched = []\n",
    "\n",
    "for f in correct_db_files:\n",
    "    \n",
    "    if f in my_db_files:\n",
    "        continue\n",
    "            \n",
    "    file_only = f.split('/')[-1]\n",
    "    \n",
    "    if '.log' in f:\n",
    "        continue\n",
    "    \n",
    "    potential_files = [i for i in my_db_files if i.split('/')[-1] == file_only]\n",
    "    if len(potential_files) > 0:\n",
    "\n",
    "        \n",
    "        if '117086' in f:\n",
    "            potential_files = [i for i in potential_files if '117086' in i]\n",
    "        elif '117088' in f:\n",
    "            potential_files = [i for i in potential_files if '117088' in i]\n",
    "        \n",
    "        print(f)\n",
    "        print('possible matches:')\n",
    "        for i in potential_files:\n",
    "            print(i)\n",
    "    \n",
    "    else:\n",
    "        print(f)\n",
    "        print('no matches found')\n",
    "    \n",
    "    print()\n",
    "    \n",
    "    not_matched+=[f]\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6723678",
   "metadata": {},
   "source": [
    "### Unmatched files:\n",
    "     \n",
    "     'pangenomes/117086/temp/vsearch/genes.ffn' - can ignore, this is the same as nextflow_db/pangenomes/117086/genes.ffn, just a different path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "17942c2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pangenomes/117088/temp/vsearch/genes.ffn',\n",
       " 'pangenomes/117086/temp/vsearch/genes.ffn']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "77661b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genes.ffn files are the same\n",
    "!diff <(sort midasdb_localdb/pangenomes/117086/temp/vsearch/genes.ffn) <(sort nextflow_db/pangenomes/117086/genes.ffn)\n",
    "!diff <(sort midasdb_localdb/pangenomes/117088/temp/vsearch/genes.ffn) <(sort nextflow_db/pangenomes/117088/genes.ffn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11dc29f3",
   "metadata": {},
   "source": [
    "### Are the contents of each file the same?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8ea55588",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ignore lines with these words\n",
    "diff_exclude = '\" -I \"'.join(['-I \"clairedubin','pollard', #names in most paths\n",
    "                        '^\\#.*', #exclude comment lines  \n",
    "                        'time', 'tm_sec', 'tm_min', 'tm_hour', 'tm_mday',\n",
    "                        'tm_mon', 'tm_wday', 'tm_yday', 'Date', #timestamp stuff\n",
    "                        'Database name', #path to databases (some aren't full paths)\n",
    "                        '\\<Statistics_',  #comments in resfinder outputs\n",
    "                             \n",
    "                             ])\n",
    "diff_exclude += '\"'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "41cf20e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#these will cause diff to fail\n",
    "binary_files = ['markers/phyeco/phyeco.fa.bwt',\n",
    "               'markers/phyeco/phyeco.fa.sa']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "faf18580",
   "metadata": {},
   "outputs": [],
   "source": [
    "different_contents = {}\n",
    "same_contents = []\n",
    "different_but_ok_files = []\n",
    "\n",
    "for f in correct_db_files:\n",
    "        \n",
    "    if f in not_matched or f in binary_files:\n",
    "        continue\n",
    "        \n",
    "    if 'temp' in f or 'tmp' in f:#skip intermediate files\n",
    "        continue\n",
    "        \n",
    "    #check if contents are the same, excluding words specified above and sorting lines\n",
    "    cmd = f\"diff {diff_exclude} <(sort nextflow_db/{f}) <(sort midasdb_localdb/{f})\"\n",
    "    diff_out = !{cmd}\n",
    "    \n",
    "    if len(diff_out) == 0:\n",
    "        same_contents += [f]\n",
    "        continue\n",
    "    \n",
    "    different_contents[f] =  [f'nextflow_db/{f}', f'midasdb_localdb/{f}', diff_out]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dd7f7031",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(344, 36)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(same_contents), len(different_contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d697efa9",
   "metadata": {},
   "source": [
    "### Which subfolders have differing files?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "90c7e6ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'pangenomes_annotation': 31, 'gene_annotations': 4, 'markers': 1})"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subs = []\n",
    "\n",
    "for i in different_contents.keys():\n",
    "    \n",
    "    subfolder = i.split('/')[0]\n",
    "    subs += [subfolder]\n",
    "    \n",
    "Counter(subs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8374f4",
   "metadata": {},
   "source": [
    "### Compare differing files in gene_annotations directory\n",
    "\n",
    "These are the same - diff just can't handle the DB files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "33cf2f5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_annotations/117086/GCF_900752885.1/GCF_900752885.1.gff.db\n",
      "gene_annotations/117086/GCA_900552055.1/GCA_900552055.1.gff.db\n",
      "gene_annotations/117088/GCA_007120565.1/GCA_007120565.1.gff.db\n",
      "gene_annotations/117088/GCA_007135585.1/GCA_007135585.1.gff.db\n"
     ]
    }
   ],
   "source": [
    "g = []\n",
    "for i in different_contents.keys():\n",
    "    if 'gene_annotations/' in i:\n",
    "        print(i)\n",
    "        g += [i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "130a8761",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in g:\n",
    "\n",
    "    test_gff = gffutils.interface.FeatureDB(f'nextflow_db/{i}')\n",
    "    correct_gff = gffutils.interface.FeatureDB(f'midasdb_localdb/{i}')\n",
    "\n",
    "    c = correct_gff.all_features()\n",
    "\n",
    "    for i in test_gff.all_features():\n",
    "        if i != next(c):\n",
    "            print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e5831a43",
   "metadata": {},
   "outputs": [],
   "source": [
    "ignore_diffs = []\n",
    "ignore_diffs += [i for i in different_contents.keys() if 'gene_annotations/' in i]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51aabfba",
   "metadata": {},
   "source": [
    "### Compare differing files in markers directory\n",
    "\n",
    "Header files are just different because of the date/path in the header."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2a7b0ee7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "markers/phyeco/phyeco.fa.header\n"
     ]
    }
   ],
   "source": [
    "for i in different_contents.keys():\n",
    "    if 'markers/' in i:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d34bc2f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = 'markers/phyeco/phyeco.fa'\n",
    "cmd = f\"diff {diff_exclude} <(sort nextflow_db/{f}) <(sort midasdb_localdb/{f})\"\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fab78db1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "92,99c92,99\n",
      "< \ttm_hour=11\n",
      "< \ttm_mday=6\n",
      "< \ttm_min=18\n",
      "< \ttm_mon=1\n",
      "< \ttm_sec=14\n",
      "< \ttm_wday=4\n",
      "< \ttm_yday=36\n",
      "< \ttm_year=125\n",
      "---\n",
      "> \ttm_hour=14\n",
      "> \ttm_mday=24\n",
      "> \ttm_min=26\n",
      "> \ttm_mon=8\n",
      "> \ttm_sec=46\n",
      "> \ttm_wday=2\n",
      "> \ttm_yday=267\n",
      "> \ttm_year=124\n"
     ]
    }
   ],
   "source": [
    "f = 'markers/phyeco/phyeco.fa.header'\n",
    "cmd = f\"diff {diff_exclude} <(sort nextflow_db/{f}) <(sort midasdb_localdb/{f})\"\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1ec85471",
   "metadata": {},
   "outputs": [],
   "source": [
    "ignore_diffs += ['markers/phyeco/phyeco.fa.header']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a2d075",
   "metadata": {},
   "source": [
    "## Compare differing files in pangenomes_annotation directory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bac660da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_aggregated_classification/GCA_007135585.1_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_aggregated_classification/GCA_007135585.1_aggregated_classification.tsv\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_marker_classification/GCA_007135585.1_features.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_score_calibration/GCA_007135585.1_calibrated_aggregated_classification.tsv\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_score_calibration/GCA_007135585.1_calibrated_nn_classification.tsv\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_score_calibration/GCA_007135585.1_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_score_calibration/GCA_007135585.1_calibrated_marker_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_score_calibration/GCA_007135585.1_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_nn_classification/GCA_007135585.1_nn_classification.tsv\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_nn_classification/GCA_007135585.1_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/genomad_output/GCA_007120565.1_aggregated_classification/GCA_007120565.1_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/genomad_output/GCA_007120565.1_nn_classification/GCA_007120565.1_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/genomad_output/GCA_007120565.1_score_calibration/GCA_007120565.1_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/genomad_output/GCA_007120565.1_score_calibration/GCA_007120565.1_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_score_calibration/GCA_900552055.1_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_score_calibration/GCA_900552055.1_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_score_calibration/GCA_900552055.1_provirus_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_score_calibration/GCA_900552055.1_provirus_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_nn_classification/GCA_900552055.1_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_nn_classification/GCA_900552055.1_provirus_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_aggregated_classification/GCA_900552055.1_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_aggregated_classification/GCA_900552055.1_provirus_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_aggregated_classification/GCF_900752885.1_provirus_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_aggregated_classification/GCF_900752885.1_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_nn_classification/GCF_900752885.1_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_nn_classification/GCF_900752885.1_provirus_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_calibrated_aggregated_classification.tsv\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_provirus_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_provirus_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_calibrated_aggregated_classification.npz\n"
     ]
    }
   ],
   "source": [
    "for i in different_contents.keys():\n",
    "    if 'pangenomes_annotation/' in i:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b95901c4",
   "metadata": {},
   "source": [
    "### Genomad outputs\n",
    "\n",
    "TSVs and .npz files are the same, except for a <0.001 rounding difference in a few scores. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a2bd28bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomad_tsvs = [i for i in different_contents.keys() if 'genomad_output' in i and i.endswith('.tsv')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3c900217",
   "metadata": {},
   "outputs": [],
   "source": [
    "for f in genomad_tsvs:\n",
    "\n",
    "    df1 = pd.read_csv(f'nextflow_db/{f}', sep='\\t', index_col=0)\n",
    "    df2 = pd.read_csv(f'midasdb_localdb/{f}', sep='\\t', index_col=0)\n",
    "\n",
    "\n",
    "    for gene, row1 in df1.iterrows():\n",
    "\n",
    "        row2 = df2.loc[gene]\n",
    "        if row1.equals(row2):\n",
    "            continue\n",
    "\n",
    "        #check if it's +/- 0.0001 difference\n",
    "        elif abs(row1-row2).sum() <= 0.001:\n",
    "            continue\n",
    "\n",
    "        else:\n",
    "            print(gene)\n",
    "            print(row1.values)\n",
    "            print(row2.values)\n",
    "            print(abs(row1-row2).sum())\n",
    "            print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "cf34e2de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n"
     ]
    }
   ],
   "source": [
    "npz_files = []\n",
    "\n",
    "for f, (test_path, correct_path, diff_out) in different_contents.items():\n",
    "\n",
    "    if '.npz' in f and 'genomad' in f:\n",
    "        npz_files += [(test_path, correct_path)]\n",
    "        continue\n",
    "\n",
    "for (f1_path, f2_path) in npz_files:\n",
    "    \n",
    "    f1 = np.load(f1_path)\n",
    "    f2 = np.load(f2_path)\n",
    "    for i in f1:\n",
    "\n",
    "        if (f1[i] == f2[i]).all():\n",
    "            continue\n",
    "\n",
    "        else:\n",
    "            diffs = np.where(abs(f1[i] - f2[i]) > 0.001)\n",
    "            if len(diffs) != 0:\n",
    "                print(diffs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b240a2c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i for i in different_contents.keys() if 'pangenomes_annotation/' in i and 'genomad' in i and not ('.npz' in i or '.tsv' in i)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5383803d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ignore_diffs += [i for i in different_contents.keys() if 'pangenomes_annotation/' in i and 'genomad' in i and ('.npz' in i or '.tsv' in i)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1631ed86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i for i in different_contents.keys() if i not in ignore_diffs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee33753a-a1e3-49c7-8171-f68c339b087a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "basics",
   "language": "python",
   "name": "basics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
