{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "115bb5c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import gffutils\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccffe49d-7787-4631-906a-709b390d4f93",
   "metadata": {},
   "source": [
    "### Unzip db builds (if necessary):\n",
    "\n",
    "    tar -zxf nextflow_db.tar.gz\n",
    "    tar -zxf midasdb_localdb.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4af8ee08",
   "metadata": {},
   "source": [
    "### Output of nextflow pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "71b0f0c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_db_path = '/wynton/home/sirota/clairedubin/MIDAS3_nextflow/testing/nextflow_db'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44904b08",
   "metadata": {},
   "source": [
    "### Output run generated Chunyu using original MIDAS scripts\n",
    "\n",
    "Original code here: https://github.com/czbiohub-sf/MIDAS/blob/master/tests/test_db.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b7abf987",
   "metadata": {},
   "outputs": [],
   "source": [
    "correct_db_path = '/wynton/home/sirota/clairedubin/MIDAS3_nextflow/testing/midasdb_localdb'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3893d82",
   "metadata": {},
   "source": [
    "### Comparing folders\n",
    "\n",
    "Mine is missing genomes.tsv, markers_models, and cleaned_imports because the nextflow pipeline doesn't require these input files to be stored within the database directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "79d32421",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chunks\tgene_annotations  markers  pangenomes  pangenomes_annotation\n"
     ]
    }
   ],
   "source": [
    "!ls {my_db_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fd14dc2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chunks\t\t gene_annotations  markers\t   pangenomes\n",
      "cleaned_imports  genomes.tsv\t   markers_models  pangenomes_annotation\n"
     ]
    }
   ],
   "source": [
    "!ls {correct_db_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b563a780",
   "metadata": {},
   "source": [
    "### Get list of files in each database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2acbd0dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = 'nextflow_db'\n",
    "my_db_files = []\n",
    "\n",
    "for path, subdirs, files in os.walk(db):\n",
    "    for name in files:\n",
    "        \n",
    "        if name == 'genomes.tsv' or 'cleaned_imports' in path:\n",
    "            continue\n",
    "        \n",
    "        my_db_files += [os.path.join(path, name).replace(db+'/', '')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6bf9a8a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = 'midasdb_localdb'\n",
    "correct_db_files = []\n",
    "\n",
    "for path, subdirs, files in os.walk(db):\n",
    "    for name in files:\n",
    "        if '.log' in name: #ignore missing log files\n",
    "            continue\n",
    "        if name == 'genomes.tsv' or 'cleaned_imports' in path or 'markers_models' in path:\n",
    "            continue\n",
    "        correct_db_files += [os.path.join(path, name).replace(db+'/', '')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab7219e9",
   "metadata": {},
   "source": [
    "### Compare list of files generated by each database\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "250bc825",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pangenomes/117088/temp/vsearch/genes.ffn\n",
      "possible matches:\n",
      "pangenomes/117088/genes.ffn\n",
      "pangenomes/117088/temp/cdhit/genes.ffn\n",
      "\n",
      "pangenomes/117086/temp/vsearch/genes.ffn\n",
      "possible matches:\n",
      "pangenomes/117086/genes.ffn\n",
      "pangenomes/117086/temp/cdhit/genes.ffn\n",
      "\n"
     ]
    }
   ],
   "source": [
    "not_matched = []\n",
    "\n",
    "for f in correct_db_files:\n",
    "    \n",
    "    if f in my_db_files:\n",
    "        continue\n",
    "            \n",
    "    file_only = f.split('/')[-1]\n",
    "    \n",
    "    if '.log' in f:\n",
    "        continue\n",
    "    \n",
    "    potential_files = [i for i in my_db_files if i.split('/')[-1] == file_only]\n",
    "    if len(potential_files) > 0:\n",
    "\n",
    "        \n",
    "        if '117086' in f:\n",
    "            potential_files = [i for i in potential_files if '117086' in i]\n",
    "        elif '117088' in f:\n",
    "            potential_files = [i for i in potential_files if '117088' in i]\n",
    "        \n",
    "        print(f)\n",
    "        print('possible matches:')\n",
    "        for i in potential_files:\n",
    "            print(i)\n",
    "    \n",
    "    else:\n",
    "        print(f)\n",
    "        print('no matches found')\n",
    "    \n",
    "    print()\n",
    "    \n",
    "    not_matched+=[f]\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6723678",
   "metadata": {},
   "source": [
    "### Unmatched files:\n",
    "     \n",
    "     'pangenomes/117086/temp/vsearch/genes.ffn' - can ignore, this is the same as nextflow_db/pangenomes/117086/genes.ffn, just a different path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "17942c2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pangenomes/117088/temp/vsearch/genes.ffn',\n",
       " 'pangenomes/117086/temp/vsearch/genes.ffn']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "77661b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genes.ffn files are the same\n",
    "!diff <(sort midasdb_localdb/pangenomes/117086/temp/vsearch/genes.ffn) <(sort nextflow_db/pangenomes/117086/genes.ffn)\n",
    "!diff <(sort midasdb_localdb/pangenomes/117088/temp/vsearch/genes.ffn) <(sort nextflow_db/pangenomes/117088/genes.ffn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11dc29f3",
   "metadata": {},
   "source": [
    "### Are the contents of each file the same?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8ea55588",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ignore lines with these words\n",
    "diff_exclude = '\" -I \"'.join(['-I \"clairedubin','pollard', #names in most paths\n",
    "                        '^\\#.*', #exclude comment lines  \n",
    "                        'time', 'tm_sec', 'tm_min', 'tm_hour', 'tm_mday',\n",
    "                        'tm_mon', 'tm_wday', 'tm_yday', 'Date', #timestamp stuff\n",
    "                        'Database name', #path to databases (some aren't full paths)\n",
    "                        '\\<Statistics_',  #comments in resfinder outputs\n",
    "                             \n",
    "                             ])\n",
    "diff_exclude += '\"'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "41cf20e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#these will cause diff to fail\n",
    "binary_files = ['markers/phyeco/phyeco.fa.bwt',\n",
    "               'markers/phyeco/phyeco.fa.sa']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "faf18580",
   "metadata": {},
   "outputs": [],
   "source": [
    "different_contents = {}\n",
    "same_contents = []\n",
    "different_but_ok_files = []\n",
    "\n",
    "for f in correct_db_files:\n",
    "        \n",
    "    if f in not_matched or f in binary_files:\n",
    "        continue\n",
    "        \n",
    "    if 'temp' in f or 'tmp' in f:#skip intermediate files\n",
    "        continue\n",
    "        \n",
    "    #check if contents are the same, excluding words specified above and sorting lines\n",
    "    cmd = f\"diff {diff_exclude} <(sort nextflow_db/{f}) <(sort midasdb_localdb/{f})\"\n",
    "    diff_out = !{cmd}\n",
    "    \n",
    "    if len(diff_out) == 0:\n",
    "        same_contents += [f]\n",
    "        continue\n",
    "    \n",
    "    different_contents[f] =  [f'nextflow_db/{f}', f'midasdb_localdb/{f}', diff_out]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dd7f7031",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(342, 38)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(same_contents), len(different_contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d697efa9",
   "metadata": {},
   "source": [
    "### Which subfolders have differing files?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "90c7e6ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'pangenomes_annotation': 33, 'gene_annotations': 4, 'markers': 1})"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subs = []\n",
    "\n",
    "for i in different_contents.keys():\n",
    "    \n",
    "    subfolder = i.split('/')[0]\n",
    "    subs += [subfolder]\n",
    "    \n",
    "Counter(subs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8374f4",
   "metadata": {},
   "source": [
    "### Compare differing files in gene_annotations directory\n",
    "\n",
    "These are the same - diff just can't handle the DB files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "33cf2f5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_annotations/117086/GCF_900752885.1/GCF_900752885.1.gff.db\n",
      "gene_annotations/117086/GCA_900552055.1/GCA_900552055.1.gff.db\n",
      "gene_annotations/117088/GCA_007120565.1/GCA_007120565.1.gff.db\n",
      "gene_annotations/117088/GCA_007135585.1/GCA_007135585.1.gff.db\n"
     ]
    }
   ],
   "source": [
    "g = []\n",
    "for i in different_contents.keys():\n",
    "    if 'gene_annotations/' in i:\n",
    "        print(i)\n",
    "        g += [i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "130a8761",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in g:\n",
    "\n",
    "    test_gff = gffutils.interface.FeatureDB(f'nextflow_db/{i}')\n",
    "    correct_gff = gffutils.interface.FeatureDB(f'midasdb_localdb/{i}')\n",
    "\n",
    "    c = correct_gff.all_features()\n",
    "\n",
    "    for i in test_gff.all_features():\n",
    "        if i != next(c):\n",
    "            print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e5831a43",
   "metadata": {},
   "outputs": [],
   "source": [
    "ignore_diffs = []\n",
    "ignore_diffs += [i for i in different_contents.keys() if 'gene_annotations/' in i]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51aabfba",
   "metadata": {},
   "source": [
    "### Compare differing files in markers directory\n",
    "\n",
    "Header files are just different because of the date/path in the header."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2a7b0ee7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "markers/phyeco/phyeco.fa.header\n"
     ]
    }
   ],
   "source": [
    "for i in different_contents.keys():\n",
    "    if 'markers/' in i:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d34bc2f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = 'markers/phyeco/phyeco.fa'\n",
    "cmd = f\"diff {diff_exclude} <(sort nextflow_db/{f}) <(sort midasdb_localdb/{f})\"\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fab78db1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "92,99c92,99\n",
      "< \ttm_hour=17\n",
      "< \ttm_mday=20\n",
      "< \ttm_min=55\n",
      "< \ttm_mon=1\n",
      "< \ttm_sec=55\n",
      "< \ttm_wday=5\n",
      "< \ttm_yday=50\n",
      "< \ttm_year=126\n",
      "---\n",
      "> \ttm_hour=14\n",
      "> \ttm_mday=24\n",
      "> \ttm_min=26\n",
      "> \ttm_mon=8\n",
      "> \ttm_sec=46\n",
      "> \ttm_wday=2\n",
      "> \ttm_yday=267\n",
      "> \ttm_year=124\n"
     ]
    }
   ],
   "source": [
    "f = 'markers/phyeco/phyeco.fa.header'\n",
    "cmd = f\"diff {diff_exclude} <(sort nextflow_db/{f}) <(sort midasdb_localdb/{f})\"\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1ec85471",
   "metadata": {},
   "outputs": [],
   "source": [
    "ignore_diffs += ['markers/phyeco/phyeco.fa.header']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a2d075",
   "metadata": {},
   "source": [
    "## Compare differing files in pangenomes_annotation directory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bac660da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/resfinder_output/pheno_table.txt\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_score_calibration/GCA_900552055.1_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_score_calibration/GCA_900552055.1_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_score_calibration/GCA_900552055.1_provirus_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_score_calibration/GCA_900552055.1_provirus_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_nn_classification/GCA_900552055.1_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_nn_classification/GCA_900552055.1_nn_classification.tsv\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_nn_classification/GCA_900552055.1_provirus_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_aggregated_classification/GCA_900552055.1_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCA_900552055.1/genomad_output/GCA_900552055.1_aggregated_classification/GCA_900552055.1_provirus_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_aggregated_classification/GCF_900752885.1_provirus_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_aggregated_classification/GCF_900752885.1_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_nn_classification/GCF_900752885.1_provirus_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_nn_classification/GCF_900752885.1_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_calibrated_aggregated_classification.tsv\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_provirus_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_provirus_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/genomad_output/GCF_900752885.1_score_calibration/GCF_900752885.1_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117086/GCF_900752885.1/resfinder_output/pheno_table.txt\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/resfinder_output/pheno_table.txt\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_nn_classification/GCA_007135585.1_nn_classification.tsv\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_nn_classification/GCA_007135585.1_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_aggregated_classification/GCA_007135585.1_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_score_calibration/GCA_007135585.1_calibrated_aggregated_classification.tsv\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_score_calibration/GCA_007135585.1_calibrated_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007135585.1/genomad_output/GCA_007135585.1_score_calibration/GCA_007135585.1_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/resfinder_output/pheno_table.txt\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/genomad_output/GCA_007120565.1_aggregated_classification/GCA_007120565.1_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/genomad_output/GCA_007120565.1_nn_classification/GCA_007120565.1_nn_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/genomad_output/GCA_007120565.1_nn_classification/GCA_007120565.1_nn_classification.tsv\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/genomad_output/GCA_007120565.1_score_calibration/GCA_007120565.1_calibrated_aggregated_classification.npz\n",
      "pangenomes_annotation/01_mge/117088/GCA_007120565.1/genomad_output/GCA_007120565.1_score_calibration/GCA_007120565.1_calibrated_nn_classification.npz\n"
     ]
    }
   ],
   "source": [
    "for i in different_contents.keys():\n",
    "    if 'pangenomes_annotation/' in i:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b95901c4",
   "metadata": {},
   "source": [
    "### Genomad outputs\n",
    "\n",
    "TSVs and .npz files are the same, except for a <0.001 rounding difference in a few scores. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a2bd28bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomad_tsvs = [i for i in different_contents.keys() if 'genomad_output' in i and i.endswith('.tsv')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3c900217",
   "metadata": {},
   "outputs": [],
   "source": [
    "for f in genomad_tsvs:\n",
    "    \n",
    "    df1 = pd.read_csv(f'nextflow_db/{f}', sep='\\t', index_col=0)\n",
    "    df2 = pd.read_csv(f'midasdb_localdb/{f}', sep='\\t', index_col=0)\n",
    "\n",
    "    numeric_cols = df1.select_dtypes(include=[np.number]).columns\n",
    "    str_cols = df1.select_dtypes(exclude=[np.number]).columns\n",
    "\n",
    "    for gene, row1 in df1.iterrows():\n",
    "        row2 = df2.loc[gene]\n",
    "        \n",
    "        if row1.equals(row2):\n",
    "            continue\n",
    "            \n",
    "\n",
    "        if not row1[str_cols].equals(row2[str_cols]):\n",
    "            print(gene)\n",
    "            print(\"String/Categorical mismatch:\")\n",
    "            print(\"Row 1:\", row1)\n",
    "            print(\"Row 2:\", row2)\n",
    "            print()\n",
    "            continue\n",
    "\n",
    "\n",
    "        numeric_diff = abs(row1[numeric_cols].fillna(0) - row2[numeric_cols].fillna(0)).sum()\n",
    "        \n",
    "        if numeric_diff <= 0.001:\n",
    "            continue\n",
    "        else:\n",
    "            print(gene)\n",
    "            print(\"Numeric mismatch:\")\n",
    "            print(\"Row 1:\", row1)\n",
    "            print(\"Row 2:\", row2)\n",
    "            print(f\"Total numeric difference: {numeric_diff}\")\n",
    "            print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "cf34e2de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n",
      "(array([], dtype=int64), array([], dtype=int64))\n"
     ]
    }
   ],
   "source": [
    "npz_files = []\n",
    "\n",
    "for f, (test_path, correct_path, diff_out) in different_contents.items():\n",
    "\n",
    "    if '.npz' in f and 'genomad' in f:\n",
    "        npz_files += [(test_path, correct_path)]\n",
    "        continue\n",
    "\n",
    "for (f1_path, f2_path) in npz_files:\n",
    "    \n",
    "    f1 = np.load(f1_path)\n",
    "    f2 = np.load(f2_path)\n",
    "    for i in f1:\n",
    "\n",
    "        if (f1[i] == f2[i]).all():\n",
    "            continue\n",
    "\n",
    "        else:\n",
    "            diffs = np.where(abs(f1[i] - f2[i]) > 0.001)\n",
    "            if len(diffs) != 0:\n",
    "                print(diffs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b240a2c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i for i in different_contents.keys() if 'pangenomes_annotation/' in i and 'genomad' in i and not ('.npz' in i or '.tsv' in i)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5383803d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ignore_diffs += [i for i in different_contents.keys() if 'pangenomes_annotation/' in i and 'genomad' in i and ('.npz' in i or '.tsv' in i)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1631ed86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pangenomes_annotation/01_mge/117086/GCA_900552055.1/resfinder_output/pheno_table.txt',\n",
       " 'pangenomes_annotation/01_mge/117086/GCF_900752885.1/resfinder_output/pheno_table.txt',\n",
       " 'pangenomes_annotation/01_mge/117088/GCA_007135585.1/resfinder_output/pheno_table.txt',\n",
       " 'pangenomes_annotation/01_mge/117088/GCA_007120565.1/resfinder_output/pheno_table.txt']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i for i in different_contents.keys() if i not in ignore_diffs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ee33753a-a1e3-49c7-8171-f68c339b087a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "68a69\n",
      "> cephalotin\tbeta-lactam\tNo resistance\t0\t\n"
     ]
    }
   ],
   "source": [
    "!diff 'nextflow_db/pangenomes_annotation/01_mge/117086/GCA_900552055.1/resfinder_output/pheno_table.txt' 'midasdb_localdb/pangenomes_annotation/01_mge/117086/GCA_900552055.1/resfinder_output/pheno_table.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "b9deee19-5d90-4167-a0b3-82266a6abba9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "68a69\n",
      "> cephalotin\tbeta-lactam\tNo resistance\t0\t\n"
     ]
    }
   ],
   "source": [
    "!diff 'nextflow_db/pangenomes_annotation/01_mge/117086/GCF_900752885.1/resfinder_output/pheno_table.txt' 'midasdb_localdb/pangenomes_annotation/01_mge/117086/GCF_900752885.1/resfinder_output/pheno_table.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a5d46f45-260e-4de6-9fe5-6f0455657a7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# ResFinder phenotype results.\n",
      "# Sample: GCA_900552055.1.fna\n",
      "# \n",
      "# The phenotype 'No resistance' should be interpreted with\n",
      "# caution, as it only means that nothing in the used\n",
      "# database indicate resistance, but resistance could exist\n",
      "# from 'unknown' or not yet implemented sources.\n",
      "# \n",
      "# The 'Match' column stores one of the integers 0, 1, 2, 3.\n",
      "#      0: No match found\n",
      "#      1: Match < 100% ID AND match length < ref length\n",
      "#      2: Match = 100% ID AND match length < ref length\n",
      "#      3: Match = 100% ID AND match length = ref length\n",
      "# If several hits causing the same resistance are found,\n",
      "# the highest number will be stored in the 'Match' column.\n",
      "\n",
      "# Antimicrobial\tClass\tWGS-predicted phenotype\tMatch\tGenetic background\n",
      "gentamicin\taminoglycoside\tNo resistance\t0\t\n",
      "tobramycin\taminoglycoside\tNo resistance\t0\t\n",
      "streptomycin\taminoglycoside\tNo resistance\t0\t\n",
      "amikacin\taminoglycoside\tNo resistance\t0\t\n",
      "isepamicin\taminoglycoside\tNo resistance\t0\t\n",
      "dibekacin\taminoglycoside\tNo resistance\t0\t\n",
      "kanamycin\taminoglycoside\tNo resistance\t0\t\n",
      "neomycin\taminoglycoside\tNo resistance\t0\t\n",
      "lividomycin\taminoglycoside\tNo resistance\t0\t\n",
      "paromomycin\taminoglycoside\tNo resistance\t0\t\n",
      "ribostamycin\taminoglycoside\tNo resistance\t0\t\n",
      "unknown aminoglycoside\taminoglycoside\tNo resistance\t0\t\n",
      "butiromycin\taminoglycoside\tNo resistance\t0\t\n",
      "butirosin\taminoglycoside\tNo resistance\t0\t\n",
      "hygromycin\taminoglycoside\tNo resistance\t0\t\n",
      "netilmicin\taminoglycoside\tNo resistance\t0\t\n",
      "apramycin\taminoglycoside\tNo resistance\t0\t\n",
      "sisomicin\taminoglycoside\tNo resistance\t0\t\n",
      "arbekacin\taminoglycoside\tNo resistance\t0\t\n",
      "kasugamycin\taminoglycoside\tNo resistance\t0\t\n",
      "astromicin\taminoglycoside\tNo resistance\t0\t\n",
      "fortimicin\taminoglycoside\tNo resistance\t0\t\n",
      "spectinomycin\taminocyclitol\tNo resistance\t0\t\n",
      "fluoroquinolone\tquinolone\tNo resistance\t0\t\n",
      "ciprofloxacin\tquinolone\tNo resistance\t0\t\n",
      "unknown quinolone\tquinolone\tNo resistance\t0\t\n",
      "nalidixic acid\tquinolone\tNo resistance\t0\t\n",
      "amoxicillin\tbeta-lactam\tNo resistance\t0\t\n",
      "amoxicillin+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "ampicillin\tbeta-lactam\tNo resistance\t0\t\n",
      "ampicillin+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "cefepime\tbeta-lactam\tNo resistance\t0\t\n",
      "cefixime\tbeta-lactam\tNo resistance\t0\t\n",
      "cefotaxime\tbeta-lactam\tNo resistance\t0\t\n",
      "cefoxitin\tbeta-lactam\tNo resistance\t0\t\n",
      "ceftazidime\tbeta-lactam\tNo resistance\t0\t\n",
      "ertapenem\tbeta-lactam\tNo resistance\t0\t\n",
      "imipenem\tbeta-lactam\tNo resistance\t0\t\n",
      "meropenem\tbeta-lactam\tNo resistance\t0\t\n",
      "piperacillin\tbeta-lactam\tNo resistance\t0\t\n",
      "piperacillin+tazobactam\tbeta-lactam\tNo resistance\t0\t\n",
      "unknown beta-lactam\tbeta-lactam\tNo resistance\t0\t\n",
      "aztreonam\tbeta-lactam\tNo resistance\t0\t\n",
      "cefotaxime+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "temocillin\tbeta-lactam\tNo resistance\t0\t\n",
      "ticarcillin\tbeta-lactam\tNo resistance\t0\t\n",
      "ceftazidime+avibactam\tbeta-lactam\tNo resistance\t0\t\n",
      "penicillin\tbeta-lactam\tNo resistance\t0\t\n",
      "ceftriaxone\tbeta-lactam\tNo resistance\t0\t\n",
      "ticarcillin+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "cephalothin\tbeta-lactam\tNo resistance\t0\t\n",
      "cephalotin\tbeta-lactam\tNo resistance\t0\t\n",
      "piperacillin+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "ceftiofur\tunder_development\tNo resistance\t0\t\n",
      "sulfamethoxazole\tfolate pathway antagonist\tNo resistance\t0\t\n",
      "trimethoprim\tfolate pathway antagonist\tNo resistance\t0\t\n",
      "fosfomycin\tfosfomycin\tNo resistance\t0\t\n",
      "vancomycin\tglycopeptide\tNo resistance\t0\t\n",
      "teicoplanin\tglycopeptide\tNo resistance\t0\t\n",
      "bleomycin\tglycopeptide\tNo resistance\t0\t\n",
      "lincomycin\tlincosamide\tNo resistance\t0\t\n",
      "clindamycin\tlincosamide\tNo resistance\t0\t\n",
      "dalfopristin\tstreptogramin a\tNo resistance\t0\t\n",
      "pristinamycin iia\tstreptogramin a\tNo resistance\t0\t\n",
      "virginiamycin m\tstreptogramin a\tNo resistance\t0\t\n",
      "quinupristin+dalfopristin\tstreptogramin a\tNo resistance\t0\t\n",
      "tiamulin\tpleuromutilin\tNo resistance\t0\t\n",
      "carbomycin\tmacrolide\tNo resistance\t0\t\n",
      "erythromycin\tmacrolide\tNo resistance\t0\t\n",
      "azithromycin\tmacrolide\tNo resistance\t0\t\n",
      "oleandomycin\tmacrolide\tNo resistance\t0\t\n",
      "spiramycin\tmacrolide\tNo resistance\t0\t\n",
      "tylosin\tmacrolide\tNo resistance\t0\t\n",
      "telithromycin\tmacrolide\tNo resistance\t0\t\n",
      "tetracycline\ttetracycline\tNo resistance\t0\t\n",
      "doxycycline\ttetracycline\tNo resistance\t0\t\n",
      "minocycline\ttetracycline\tNo resistance\t0\t\n",
      "tigecycline\ttetracycline\tNo resistance\t0\t\n",
      "quinupristin\tstreptogramin b\tNo resistance\t0\t\n",
      "pristinamycin ia\tstreptogramin b\tNo resistance\t0\t\n",
      "virginiamycin s\tstreptogramin b\tNo resistance\t0\t\n",
      "linezolid\toxazolidinone\tNo resistance\t0\t\n",
      "chloramphenicol\tamphenicol\tNo resistance\t0\t\n",
      "florfenicol\tamphenicol\tNo resistance\t0\t\n",
      "colistin\tpolymyxin\tNo resistance\t0\t\n",
      "fusidic acid\tsteroid antibacterial\tNo resistance\t0\t\n",
      "mupirocin\tpseudomonic acid\tNo resistance\t0\t\n",
      "rifampicin\trifamycin\tNo resistance\t0\t\n",
      "metronidazole\tnitroimidazole\tNo resistance\t0\t\n",
      "narasin\tionophores\tNo resistance\t0\t\n",
      "salinomycin\tionophores\tNo resistance\t0\t\n",
      "maduramicin\tionophores\tNo resistance\t0\t\n",
      "formaldehyde\taldehyde\tNo resistance\t0\t\n",
      "benzylkonium chloride\tquaternary ammonium compound\tNo resistance\t0\t\n",
      "ethidium bromide\tquaternary ammonium compound\tNo resistance\t0\t\n",
      "chlorhexidine\tquaternary ammonium compound\tNo resistance\t0\t\n",
      "cetylpyridinium chloride\tquaternary ammonium compound\tNo resistance\t0\t\n",
      "hydrogen peroxide\tperoxide\tNo resistance\t0\t\n",
      "temperature\theat\tNo resistance\t0\t\n"
     ]
    }
   ],
   "source": [
    "!cat 'midasdb_localdb/pangenomes_annotation/01_mge/117086/GCA_900552055.1/resfinder_output/pheno_table.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "3bb9b91b-1138-4f37-9067-330c0a3d5e3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "disinfinder_blast\t\t     pheno_table.txt\n",
      "DisinFinder_Hit_in_genome_seq.fsa    resfinder_blast\n",
      "DisinFinder_Resistance_gene_seq.fsa  ResFinder_Hit_in_genome_seq.fsa\n",
      "DisinFinder_results_table.txt\t     ResFinder_Resistance_gene_seq.fsa\n",
      "DisinFinder_results_tab.txt\t     ResFinder_results_table.txt\n",
      "DisinFinder_results.txt\t\t     ResFinder_results_tab.txt\n",
      "GCA_900552055.json\t\t     ResFinder_results.txt\n"
     ]
    }
   ],
   "source": [
    "! midasdb_localdb/pangenomes_annotation/01_mge/117086/GCA_900552055.1/resfinder_output/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "7e45a0e7-d26a-43b5-9998-8342cf50d428",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Disinfectants\n",
      "No hit found\n",
      "\n",
      "Resistance gene\tIdentity\tAlignment Length/Gene Length\tCoverage\tPosition in reference\tContig\tPosition in contig\tPhenotype\tAccession no.\n",
      "# ResFinder phenotype results.\n",
      "# Sample: GCA_900552055.1.fna\n",
      "# \n",
      "# The phenotype 'No resistance' should be interpreted with\n",
      "# caution, as it only means that nothing in the used\n",
      "# database indicate resistance, but resistance could exist\n",
      "# from 'unknown' or not yet implemented sources.\n",
      "# \n",
      "# The 'Match' column stores one of the integers 0, 1, 2, 3.\n",
      "#      0: No match found\n",
      "#      1: Match < 100% ID AND match length < ref length\n",
      "#      2: Match = 100% ID AND match length < ref length\n",
      "#      3: Match = 100% ID AND match length = ref length\n",
      "# If several hits causing the same resistance are found,\n",
      "# the highest number will be stored in the 'Match' column.\n",
      "\n",
      "# Antimicrobial\tClass\tWGS-predicted phenotype\tMatch\tGenetic background\n",
      "gentamicin\taminoglycoside\tNo resistance\t0\t\n",
      "tobramycin\taminoglycoside\tNo resistance\t0\t\n",
      "streptomycin\taminoglycoside\tNo resistance\t0\t\n",
      "amikacin\taminoglycoside\tNo resistance\t0\t\n",
      "isepamicin\taminoglycoside\tNo resistance\t0\t\n",
      "dibekacin\taminoglycoside\tNo resistance\t0\t\n",
      "kanamycin\taminoglycoside\tNo resistance\t0\t\n",
      "neomycin\taminoglycoside\tNo resistance\t0\t\n",
      "lividomycin\taminoglycoside\tNo resistance\t0\t\n",
      "paromomycin\taminoglycoside\tNo resistance\t0\t\n",
      "ribostamycin\taminoglycoside\tNo resistance\t0\t\n",
      "unknown aminoglycoside\taminoglycoside\tNo resistance\t0\t\n",
      "butiromycin\taminoglycoside\tNo resistance\t0\t\n",
      "butirosin\taminoglycoside\tNo resistance\t0\t\n",
      "hygromycin\taminoglycoside\tNo resistance\t0\t\n",
      "netilmicin\taminoglycoside\tNo resistance\t0\t\n",
      "apramycin\taminoglycoside\tNo resistance\t0\t\n",
      "sisomicin\taminoglycoside\tNo resistance\t0\t\n",
      "arbekacin\taminoglycoside\tNo resistance\t0\t\n",
      "kasugamycin\taminoglycoside\tNo resistance\t0\t\n",
      "astromicin\taminoglycoside\tNo resistance\t0\t\n",
      "fortimicin\taminoglycoside\tNo resistance\t0\t\n",
      "spectinomycin\taminocyclitol\tNo resistance\t0\t\n",
      "fluoroquinolone\tquinolone\tNo resistance\t0\t\n",
      "ciprofloxacin\tquinolone\tNo resistance\t0\t\n",
      "unknown quinolone\tquinolone\tNo resistance\t0\t\n",
      "nalidixic acid\tquinolone\tNo resistance\t0\t\n",
      "amoxicillin\tbeta-lactam\tNo resistance\t0\t\n",
      "amoxicillin+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "ampicillin\tbeta-lactam\tNo resistance\t0\t\n",
      "ampicillin+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "cefepime\tbeta-lactam\tNo resistance\t0\t\n",
      "cefixime\tbeta-lactam\tNo resistance\t0\t\n",
      "cefotaxime\tbeta-lactam\tNo resistance\t0\t\n",
      "cefoxitin\tbeta-lactam\tNo resistance\t0\t\n",
      "ceftazidime\tbeta-lactam\tNo resistance\t0\t\n",
      "ertapenem\tbeta-lactam\tNo resistance\t0\t\n",
      "imipenem\tbeta-lactam\tNo resistance\t0\t\n",
      "meropenem\tbeta-lactam\tNo resistance\t0\t\n",
      "piperacillin\tbeta-lactam\tNo resistance\t0\t\n",
      "piperacillin+tazobactam\tbeta-lactam\tNo resistance\t0\t\n",
      "unknown beta-lactam\tbeta-lactam\tNo resistance\t0\t\n",
      "aztreonam\tbeta-lactam\tNo resistance\t0\t\n",
      "cefotaxime+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "temocillin\tbeta-lactam\tNo resistance\t0\t\n",
      "ticarcillin\tbeta-lactam\tNo resistance\t0\t\n",
      "ceftazidime+avibactam\tbeta-lactam\tNo resistance\t0\t\n",
      "penicillin\tbeta-lactam\tNo resistance\t0\t\n",
      "ceftriaxone\tbeta-lactam\tNo resistance\t0\t\n",
      "ticarcillin+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "cephalothin\tbeta-lactam\tNo resistance\t0\t\n",
      "cephalotin\tbeta-lactam\tNo resistance\t0\t\n",
      "piperacillin+clavulanic acid\tbeta-lactam\tNo resistance\t0\t\n",
      "ceftiofur\tunder_development\tNo resistance\t0\t\n",
      "sulfamethoxazole\tfolate pathway antagonist\tNo resistance\t0\t\n",
      "trimethoprim\tfolate pathway antagonist\tNo resistance\t0\t\n",
      "fosfomycin\tfosfomycin\tNo resistance\t0\t\n",
      "vancomycin\tglycopeptide\tNo resistance\t0\t\n",
      "teicoplanin\tglycopeptide\tNo resistance\t0\t\n",
      "bleomycin\tglycopeptide\tNo resistance\t0\t\n",
      "lincomycin\tlincosamide\tNo resistance\t0\t\n",
      "clindamycin\tlincosamide\tNo resistance\t0\t\n",
      "dalfopristin\tstreptogramin a\tNo resistance\t0\t\n",
      "pristinamycin iia\tstreptogramin a\tNo resistance\t0\t\n",
      "virginiamycin m\tstreptogramin a\tNo resistance\t0\t\n",
      "quinupristin+dalfopristin\tstreptogramin a\tNo resistance\t0\t\n",
      "tiamulin\tpleuromutilin\tNo resistance\t0\t\n",
      "carbomycin\tmacrolide\tNo resistance\t0\t\n",
      "erythromycin\tmacrolide\tNo resistance\t0\t\n",
      "azithromycin\tmacrolide\tNo resistance\t0\t\n",
      "oleandomycin\tmacrolide\tNo resistance\t0\t\n",
      "spiramycin\tmacrolide\tNo resistance\t0\t\n",
      "tylosin\tmacrolide\tNo resistance\t0\t\n",
      "telithromycin\tmacrolide\tNo resistance\t0\t\n",
      "tetracycline\ttetracycline\tNo resistance\t0\t\n",
      "doxycycline\ttetracycline\tNo resistance\t0\t\n",
      "minocycline\ttetracycline\tNo resistance\t0\t\n",
      "tigecycline\ttetracycline\tNo resistance\t0\t\n",
      "quinupristin\tstreptogramin b\tNo resistance\t0\t\n",
      "pristinamycin ia\tstreptogramin b\tNo resistance\t0\t\n",
      "virginiamycin s\tstreptogramin b\tNo resistance\t0\t\n",
      "linezolid\toxazolidinone\tNo resistance\t0\t\n",
      "chloramphenicol\tamphenicol\tNo resistance\t0\t\n",
      "florfenicol\tamphenicol\tNo resistance\t0\t\n",
      "colistin\tpolymyxin\tNo resistance\t0\t\n",
      "fusidic acid\tsteroid antibacterial\tNo resistance\t0\t\n",
      "mupirocin\tpseudomonic acid\tNo resistance\t0\t\n",
      "rifampicin\trifamycin\tNo resistance\t0\t\n",
      "metronidazole\tnitroimidazole\tNo resistance\t0\t\n",
      "narasin\tionophores\tNo resistance\t0\t\n",
      "salinomycin\tionophores\tNo resistance\t0\t\n",
      "maduramicin\tionophores\tNo resistance\t0\t\n",
      "formaldehyde\taldehyde\tNo resistance\t0\t\n",
      "benzylkonium chloride\tquaternary ammonium compound\tNo resistance\t0\t\n",
      "ethidium bromide\tquaternary ammonium compound\tNo resistance\t0\t\n",
      "chlorhexidine\tquaternary ammonium compound\tNo resistance\t0\t\n",
      "cetylpyridinium chloride\tquaternary ammonium compound\tNo resistance\t0\t\n",
      "hydrogen peroxide\tperoxide\tNo resistance\t0\t\n",
      "temperature\theat\tNo resistance\t0\t\n",
      "Aminoglycoside\n",
      "No hit found\n",
      "\n",
      "Beta-lactam\n",
      "No hit found\n",
      "\n",
      "Colistin\n",
      "No hit found\n",
      "\n",
      "Fosfomycin\n",
      "No hit found\n",
      "\n",
      "Fusidic Acid\n",
      "No hit found\n",
      "\n",
      "MLS - Macrolide, Lincosamide and Streptogramin B\n",
      "No hit found\n",
      "\n",
      "Misc\n",
      "No hit found\n",
      "\n",
      "Nitroimidazole\n",
      "No hit found\n",
      "\n",
      "Oxazolidinone\n",
      "No hit found\n",
      "\n",
      "Phenicol\n",
      "No hit found\n",
      "\n",
      "Fluoroquinolone\n",
      "No hit found\n",
      "\n",
      "Rifampicin\n",
      "No hit found\n",
      "\n",
      "Sulphonamide\n",
      "No hit found\n",
      "\n",
      "Tetracycline\n",
      "No hit found\n",
      "\n",
      "Trimethoprim\n",
      "No hit found\n",
      "\n",
      "Glycopeptide\n",
      "No hit found\n",
      "\n",
      "Pseudomonic Acid\n",
      "No hit found\n",
      "\n",
      "Resistance gene\tIdentity\tAlignment Length/Gene Length\tCoverage\tPosition in reference\tContig\tPosition in contig\tPhenotype\tAccession no.\n"
     ]
    }
   ],
   "source": [
    "!cat midasdb_localdb/pangenomes_annotation/01_mge/117086/GCA_900552055.1/resfinder_output/*txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92e9e45f-4688-4b82-afce-c6c33e252f83",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "basics",
   "language": "python",
   "name": "basics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
